lib/Gateway/HttpClient/Stream.php

```diff
@@ -1,178 +1,154 @@
-     */
+    /** @var int */
     private $connectionTimeout;
 
-    /**
-     * @var int
-     */
+    /** @var int */
     private $connectionRetry;
 
-    /**
-     * @var int
-     */
+    /** @var int */
     private $retryWaitMs;
 
+    /** @var \Symfony\Contracts\HttpClient\HttpClientInterface */
+    private $client;
+
     /**
      * Stream constructor.
      *
      * @param int $timeout Timeout for connection in seconds.
      * @param int $retry Number of times to re-try connection.
-     * @param int $retryWaitMs Time in milli seconds.
+     * @param int $retryWaitMs Time in milliseconds.
      */
-    public function __construct(LoggerInterface $logger = null, $timeout = 10, $retry = 5, $retryWaitMs = 100)
-    {
-        $this->logger = $logger;
+    public function __construct(
+        HttpClientInterface $client,
+        int $timeout = 10,
+        int $retry = 5,
+        int $retryWaitMs = 100
+    ) {
+        $this->client = $client;
         $this->connectionTimeout = $timeout;
         $this->connectionRetry = $retry;
         $this->retryWaitMs = $retryWaitMs;
+        $this->setLogger(new NullLogger());
     }
 
     /**
-     * Execute a HTTP request to the remote server.
+     * Execute an HTTP request to the remote server.
      *
      * Returns the result from the remote server.
      *
      * @param string $method
      * @param string $path
-     * @param Message $message
      *
-     * @return Message
+     * @throws \Symfony\Contracts\HttpClient\Exception\ServerExceptionInterface
+     * @throws \Symfony\Contracts\HttpClient\Exception\ClientExceptionInterface
+     * @throws \Symfony\Contracts\HttpClient\Exception\RedirectionExceptionInterface
      */
-    public function request($method, Endpoint $endpoint, $path, Message $message = null)
+    public function request($method, Endpoint $endpoint, $path, Message $message = null): Message
     {
-        $message = $message ?: new Message();
+        $message = $message ?? new Message();
 
         // We'll try to reach backend several times before throwing exception.
         $i = 0;
         do {
             ++$i;
-            if ($responseMessage = $this->requestStream($method, $endpoint, $path, $message)) {
-                return $responseMessage;
+            try {
+                if (
+                    $responseMessage = $this->getResponseMessage(
+                        $method,
+                        $endpoint,
+                        $path,
+                        $message
+                    )
+                ) {
+                    return $responseMessage;
+                }
+            } catch (TransportExceptionInterface $e) {
+                $this->logger->warning(
+                    sprintf(
+                        'Connection attempt #%d to %s failed, retrying after %d ms',
+                        $i,
+                        $endpoint->getURL(),
+                        $this->retryWaitMs
+                    )
```

Comment:
Pass exception to logger context.

```suggestion
                $this->logger->warning(
                    sprintf(
                        'Connection attempt #%d to %s failed, retrying after %d ms',
                        $i,
                        $endpoint->getURL(),
                        $this->retryWaitMs
                    ), [
                        'exception' => $e,
                    ]
```